homeassistant/components/dwd_weather_warnings/sensor.py

```diff
@@ -65,190 +63,110 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        else:
+            data["last_update"] = None
 
-        if self._var_id == "current_warning_level":
-            prefix = "current"
-        elif self._var_id == "advance_warning_level":
-            prefix = "advance"
+        if self._sensor_type == "current_warning_level":
+            searched_warnings = self._api.current_warnings
         else:
-            raise Exception("Unknown warning type")
+            searched_warnings = self._api.expected_warnings
 
-        data["warning_count"] = self._api.data[f"{prefix}_warning_count"]
+        data["warning_count"] = len(searched_warnings)
         i = 0
-        for event in self._api.data[f"{prefix}_warnings"]:
+        for warning in searched_warnings:
             i = i + 1
-
-            # dictionary for the attribute containing the complete warning as json
-            event_json = event.copy()
-
-            data[f"warning_{i}_name"] = event["event"]
-            data[f"warning_{i}_level"] = event["level"]
-            data[f"warning_{i}_type"] = event["type"]
-            if event["headline"]:
-                data[f"warning_{i}_headline"] = event["headline"]
-            if event["description"]:
-                data[f"warning_{i}_description"] = event["description"]
-            if event["instruction"]:
-                data[f"warning_{i}_instruction"] = event["instruction"]
-
-            if event["start"] is not None:
-                data[f"warning_{i}_start"] = dt_util.as_local(
-                    dt_util.utc_from_timestamp(event["start"] / 1000)
-                )
-                event_json["start"] = data[f"warning_{i}_start"]
-
-            if event["end"] is not None:
-                data[f"warning_{i}_end"] = dt_util.as_local(
-                    dt_util.utc_from_timestamp(event["end"] / 1000)
-                )
-                event_json["end"] = data[f"warning_{i}_end"]
-
-            data[f"warning_{i}"] = event_json
+            data[f"warning_{i}_name"] = warning["event"]
+            data[f"warning_{i}_type"] = warning["event_code"]
+            data[f"warning_{i}_level"] = warning["level"]
+            data[f"warning_{i}_headline"] = warning["headline"]
+            data[f"warning_{i}_description"] = warning["description"]
+            data[f"warning_{i}_instruction"] = warning["instruction"]
+            if warning["start_time"] is not None:
+                data[f"warning_{i}_start"] = dt_util.as_local(warning["start_time"])
+            else:
+                data[f"warning_{i}_start"] = None
+            if warning["end_time"] is not None:
+                data[f"warning_{i}_end"] = dt_util.as_local(warning["end_time"])
+            else:
+                data[f"warning_{i}_end"] = None
+            data[f"warning_{i}_parameters"] = warning["parameters"]
+            data[f"warning_{i}_color"] = warning["color"]
+
+            # Dictionary for the attribute containing the complete warning
+            warning_copy = warning.copy()
```

Comment:
What's going on here?

Reply:
The `warning.copy()` came in lately with pull request https://github.com/home-assistant/core/pull/29005
I just keept it.